walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,69 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Hashtable;
+import java.util.Set;
+import java.util.List;
+import java.lang.*;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+    //Request information
+    long requestDuration = request.getDuration();
+    Collection<String> requestAttendees = request.getAttendees();
+
+    // Traverse collection of events to get attendees, and time range of each relevant event
+    for(Event event: events){
+        TimeRange eventRange = event.getWhen();
+
+        // Modify copy of Event Attendees to only include relevant attendees that overlap with the request
+        Set<String> eventAttendees = event.getAttendees();
+        Set<String> copy = new HashSet<>();
+        copy.addAll(eventAttendees);
+        copy.retainAll(requestAttendees);
+        
+        // If any of the attendees overlap with this event's attendees, then this duration is unavailable  
+        if (!(copy.isEmpty())){
+            meetingTimes.add(eventRange);
+        }
+    }
+
+    // Convert Collection of TimeRanges into a List in order to sort in ascending order
+    List<TimeRange> listMeetingTimes = new ArrayList<TimeRange>(meetingTimes);
+    Collections.sort(listMeetingTimes, TimeRange.ORDER_BY_START);
+
+    // Final Collection to hold available time ranges for the requested meeting
+    Collection<TimeRange> availableTimes = new ArrayList<TimeRange>();
+
+    int start = TimeRange.START_OF_DAY;
+    int end = TimeRange.END_OF_DAY;
+    
+    for(TimeRange range: listMeetingTimes){
+        // Create time range in between scheduled times if there is enough time allocated
+        if((range.start() > start) && ((range.start() - start) >= requestDuration)){
+            TimeRange timeTemp = TimeRange.fromStartEnd(start, range.start(), false);
```

Comment:
Here and elsewhere: We should document the boolean argument to make it easier for the code reader to understand what it is used for.
Here is the way I do it:
`TimeRange.fromStartEnd(start, range.start(), /* inclusive= */ false)